dask_planner/src/sql/optimizer/dynamic_partition_pruning.rs

```diff
@@ -0,0 +1,968 @@
+//! Optimizer rule for dynamic partition pruning (DPP)
+//!
+//! DPP refers to a query optimization rule in which distinct values in an inner join are used as
+//! filters in a table scan. This allows us to eliminate all other rows which do not fit the join
+//! condition from being read at all.
+//!
+//! Furthermore, a table involved in a join may be filtered during a scan, which allows us to
+//! further prune the values to be read.
+
+use std::{
+    collections::{HashMap, HashSet},
+    fs,
+    hash::{Hash, Hasher},
+};
+
+use datafusion_python::{
+    datafusion::parquet::{
+        basic::Type as BasicType,
+        file::reader::{FileReader, SerializedFileReader},
+        record::{reader::RowIter, RowAccessor},
+        schema::{parser::parse_message_type, types::Type},
+    },
+    datafusion_common::{Column, Result, ScalarValue},
+    datafusion_expr::{
+        logical_plan::LogicalPlan,
+        utils::from_plan,
+        Expr,
+        JoinType,
+        Operator,
+        TableScan,
+    },
+    datafusion_optimizer::{OptimizerConfig, OptimizerRule},
+};
+use log::warn;
+
+use crate::sql::table::DaskTableSource;
+
+// Optimizer rule for dynamic partition pruning
+pub struct DynamicPartitionPruning {}
+
+impl DynamicPartitionPruning {
+    #[allow(missing_docs)]
+    #[allow(dead_code)]
+    pub fn new() -> Self {
+        Self {}
+    }
+}
+
+impl OptimizerRule for DynamicPartitionPruning {
+    fn name(&self) -> &str {
+        "dynamic_partition_pruning"
+    }
+
+    fn try_optimize(
+        &self,
+        plan: &LogicalPlan,
+        _config: &dyn OptimizerConfig,
+    ) -> Result<Option<LogicalPlan>> {
+        // Parse the LogicalPlan and store tables and columns being (inner) joined upon. We do this
+        // by creating a HashSet of all InnerJoins' join.on and join.filters
+        let join_conds = gather_joins(plan);
+        let tables = gather_tables(plan);
+        let aliases = gather_aliases(plan);
+
+        if join_conds.is_empty() || tables.is_empty() {
+            // No InnerJoins to optimize with
+            Ok(None)
```

Comment:
I'm probably looking at this wrong but wouldn't this cause the `optimized_plan` variable in `optimizer.rs` to end up as `None`? Maybe the syntax defaults back to the output from `.optimize(existing_plan.original_plan)`?? I honestly don't know? More of a question than a statement.

Reply:
From my understanding, returning `Ok(None)` is the same as returning `Ok(plan.clone())`. At least that's what [this comment](https://github.com/apache/arrow-datafusion/blob/beee1d91303d5eff220fadd08b2c28404c2b3e5a/datafusion/optimizer/src/optimizer.rs#L58) indicates, I haven't actually dug too deep into the optimizer code on the DataFusion side.